### PR TITLE
feat(config): Add script for extracting/inserting secret keys from SOPS config

### PR DESCRIPTION
**Pull request by rfk**
Here's my initial attempt at #1897.  Given a `.yaml` file encrypted with SOPS, I can do the following to extract the various OpenID secret keys to local files:

```
./scripts/sops-key-config.js extract path/to/config.yaml
```

Then I can operate on them locally using `prepare-new-signing-key.ks`, `activate-new-signing-key.js`, etc. Finally I can write them back to the encrypted SOPS file via:

```
./scripts/sops-key-config.js insert path/to/config.yaml
```

The resulting SOPS file looks like:

```
openid:
    key:
        kty: ENC[AES256_GCM,data:TGCb,iv:Wjymx0Ev5XWmPJ8reqv2KNyhihL/ODQJeHRLIdxzXJs=,tag:wNcIYXGHiun0kXP8C8Mrmg==,type:str]
        "n": ENC[AES256_GCM,data:cVEKFlw50kLKdZvPMec6I4aMhy2nCly7iwspcgEvfjoB77iTjdBIbjdr/OyFs/U/I9xU1E5wBwEHygvaaY66nFWohC0Eo6OuuHcx4Kvq7zonAlKBWIqBWVLr60jU5QwI5D/0F6sWQwcSNbYlBiOX5GgaEQnRu9UOu+wCZTvod947pjO7v/ljAVxj54ibpHW8Z3jVOCtS6BsJDq/vt+aEkFsy+WUHa394Lcs3TjPL2QBENhsRtV4Uy1Ih/s2PCdtUTrT+EzYQR2wdqnKtDlg6AW9RPE3FDF3QJrpLd/308RABnzSJMIOfYdlDyhxB/S+5EJMFyoyQSsgn6TxOZn2prMHBy4jdMJmBbSXppXhgdwj527fonU0TrK/ZEJan5pOkyTaQ/ti9ctQvkp3/ovcbiLysrXZD2v2FAcdf9C+baMlAGeSYGS70TDJG/zpRzJJwbXiXQP3f,iv:A+0ucIQjXXPlUMK18DgRkVAbmCNXFOD1f/l/nTfJduA=,tag:A0xD9M0YO1z6xJ1R5rrUxQ==,type:str]
        e: ENC[AES256_GCM,data:ZoqVBQ==,iv:DW0xQ8g62aiXuKi8vdcuO0ReIQR9Wd65PinTQ7VWFe4=,tag:RH1oPrnQ35+51+xojIFh4Q==,type:str]
        d: ENC[AES256_GCM,data:h9YcQ0n4p1/P/gworjGa/gZwqwdhxOY3ihFRpZpZByJ0o6XjO0IJPiPD8TBzSnUYlQrnkjfkD+6As2Qcn/RGxyQ8tI8oKluPib7tZwCVJ7xsT/PFmxQJm/OMaNnipR8uahM44vtqgevXcFTIt4e0qnEF8DvXcI/BcxAbLPAlm5RtLibVKXdlb3GRQIND40U7fHZvhK9pI1zaKaMTtFRYHLTUU4L0+lj3TjHqX2VlMTCRSRYxyUVE3GUm8t5UUQEIrBzdvvh380bWJHChoD38/eU+zDwrju0O8RNT7lpd3mP2Ch+Ba3GiHkfJESvQKj9GY/w+uc/ix+Mizfx4Hc7g7XItTvn4Gqo1myVQ4nb+e0AJBaBbk4YEpR975xQQ4oyjpc6mLY7KUcINMpwy2boZBgZa1vl4yGOdVAPsXMWRSweGRe1jqSdmdauriZOhizZBjOsS8OL5,iv:NDNrHsXZEXIlhp6de9uTIEO2NIj+r+KgomHtJ8exSJ0=,tag:d2ZuSmFGuN7PF1WzOd7ftg==,type:str]
        p: ENC[AES256_GCM,data:vR/HNiMZwCUcPJgw6QrBDxqVyvafASpEp9H8GVvfVOggzYmUQX1Il4hj+yHu6AM+Bg94fEz5XHd3XHilLDCdjzbTAy4azT/R7rsDb6UZ+M3yL6w5foKE3raT9qZYfyE8EFXOGFL4J9th4QYgHoOIWziX3UIYPUZtQ9eSIecvi5Z7o8Pt9qpPpaeNBhwBHIlD2gNQKNOEzCLuO3/Ad2db1kCD6aF1hhebl3Ua,iv:Q40PEa7C0wCjf3rVFBX9NaTaPXYeCuXtdRfmMlqwAv0=,tag:9/Zphc8m/R2xVVDoG39MhA==,type:str] 
        q: ENC[AES256_GCM,data:t6IN1a/0/Wws+al0Appji4Gl8n1LZrlizAFFEIcUL7Mb9UyyTWwi0j/NlvepmT3NN0DRgooydfdvYOj0T5GJ9qyGuf6mwEpLxX7T2WoPaDUlMoLgEj4OXmazWp9wtlymC1ICcwRJoWn/7Jg41NfI0jZjqORDZfI9h5obJA4L9xavsShfTC83gttnLzI1ZO7sOjsiZrYOvvA975uS7gV9AwyXT9+LoLmU5+z6,iv:5B33mGdTZbSux7YK2Wu30X0ithjewAJsL58+swDr/7U=,tag:pce0M4WcQew220RABwEwRQ==,type:str] 
        dp: ENC[AES256_GCM,data:OOEWi5jBH3HtCq2g7FM/xlzJ2do/vdL+qd6T9hnO+7LDKOYT3K9C06p2QWa0QqgqzPl9CbMDWnmCDX2YY1DY5aVyE7wA/DOZzBZv71PbkG4Wwzj4QJ3Jcq8va8wM3SyNByaE/t9uAqxwB3QuRIOkXavGZX/vNXgAbPdcUL337OYvoTkFsLf3ogYyrF9vNOUCa1DfZXU3A52dYxYOv93WN19dNQcNxKMamvrE,iv:mOpYbPvBlgPBjw4NS+R+0bfoR8aN/BnBmfQHCGPo7OI=,tag:HwGlKBIwozY0AhSe5adqRQ==,type:str]
        dq: ENC[AES256_GCM,data:ZIDOaqXofMkhFG+6oktGDdZEqVDGpsgLcGVg4cfWb4rydAOfZq1Akg20KL3YzWFQUNcaInQf90GtfYUd0hVQvollvv4GA9YqG4jt8EjqsybRQB4NEtWA4XaGrXutv8Md6JYcv41Mmtr6uGSCb9qj+LEVbh7CWhcnaEzZsj4Owne6EI+mVsb4KGygzXwvGYAKizXQfQBzH1zlWOW3sK6NCVhfXYeuCbXTPVRR,iv:XG06PKBKUYN0EvVtoCDVn8sguSoXHJZKEkf0G/AA+b0=,tag:iN0LyeFs5UifdYlUm8Y2OA==,type:str]
        qi: ENC[AES256_GCM,data:yGOQBTv6Xk0DsFEjdw1yN7krw62hOw94Ysa/ijmKSQ5x9u8DBMT7tFmtar6b2VRZ2NPGa6vld4+IQKOr7jMgNuIdFlqoHOU95W6evSUUCZlIKRxU4PZp6/Mjq9MvR5+XAC7+e0Ej5ChGxaIfeFv+cnlKZ3WDYt1XbRVW16+fI0Wxl1/j9ZkWhX8EFz5+oHOHpw9g0lH6ruRIjFkN2KBqt+B151t9PO8E8sO4,iv:N3g9sb1PlBsMz6ppbO+KMelPf/USdtbBKAuELCM4qm4=,tag:WOE0mHjbiQC6+ZJNrrzQ9A==,type:str]
        kid: ENC[AES256_GCM,data:qFYZPtS/jik+FpA6Fea9Jdw=,iv:PjQJZMiUp19cSPzabhgeUy6+4fPW4wuPVbHD7uvhLWk=,tag:Nfa9hxTkFjMA7ymkZsxFyA==,type:str]
        alg: ENC[AES256_GCM,data:tMztG+g=,iv:F20V/lnkkhodMrWRYBPP+B7whG3Yha1P5MkUeJ2UDGk=,tag:jo3cQkUe5KrIJiQz7faiOg==,type:str]
        use: ENC[AES256_GCM,data:/4LZ,iv:lprVlWHZ3sx2GdkAakgClCgZpEhjMSmE67MuA5Ivz+w=,tag:Okz280Bi5hJMOcr9BPu7UQ==,type:str]
        fxa-createdAt: ENC[AES256_GCM,data:78RVBzFgf6/13Q==,iv:dbuGETEUgcaLOGShrZAgj6HANwTa8tfkpvNzJ+zJdZM=,tag:XVy2T6mb1Usy0rrj+VqtGQ==,type:float]
    newKey: null
    oldKey:
        kty: ENC[AES256_GCM,data:teZu,iv:8b/FSZHn+xtWchz/r3yEHcv2RqK08cpRhkC1R1uymks=,tag:URFPEW962NkQD4YLEAaBiw==,type:str]
        alg: ENC[AES256_GCM,data:dS5jAJs=,iv:TOrHA19aeDwCF/U5besbGishVVk3qUsEoz2uTC9v0dQ=,tag:jFHAjwRkVjtI3ZcpxFCjSA==,type:str]
        kid: ENC[AES256_GCM,data:Epuv+6nfO2r9jcHClrHyX7k=,iv:i4gOTw2vRofShp/8+zsI+kGCK1oRk/al14cNw8ICp5k=,tag:Nm/4yI9OZ+EssyBV0N0fCw==,type:str]
        fxa-createdAt: ENC[AES256_GCM,data:bXkD99gQ99Smgw==,iv:2h5h7pE5IjyG5K4LFjWnmFO5QHaQ6uZHYZJNbNgC5x0=,tag:KS7zQxRSBAX6TXiTMeR39Q==,type:float]
        use: ENC[AES256_GCM,data:Wuwo,iv:vphbdg1sDzvIH5e2MxVe84awJCha+bX0VN1ItYsnct0=,tag:0mprk2blliInaZEcdSe+Zw==,type:str]
        "n": ENC[AES256_GCM,data:GKWZIG0FOrL97zn2Q3UbygccpGGFSAPsUu9gassRhJRSJ6VosJX/cLLHVeQCM+0DsqJpS9Cti1b6nJn3Y/y94ABI/g0KrwLs66Q526zYb3aM4WK3FdW6IRwbkTOEzmhglYSLC7M9zah4Jrp76/vNw0iipT+bKhrKX825PTTuou0cK5L5wtmTh5WX8OzUzjATe6yQVe9xFd0im9/aS1eRu63qa36t2a0v0kuzS4/NqGdgHlUUJ6MefVXolhVgW6ubE1zW3XnwdvKDhK4Jtw8UIOS7i5WqrSVWkSFVijAFfo5/Hy0Q/5SXgoHm30f3ZRgG/StQEetEnEnhAjrKP87ELaj1iJAMbG/oq6czAvL6dG3DQGOBxdgZtRd4EN+NRElrUSMqDCD2vvTcEbC2suuikwjq0OpRz03DdEaQwj6YjHTy3KhrVSYC3XgpBo5GRHz7/RuJH1gi,iv:SkGEjK5khVStCPcdD9RyLQIYSLQheqHzkUw++L/nUeQ=,tag:QtPKnkfIvqbdgrX/u9PPgw==,type:str]
        e: ENC[AES256_GCM,data:KNAtNw==,iv:y/Pvjpxd8WS7KrhDYNS2kTh+PEsZeKW3WwInVi7EVAc=,tag:GqR+g6+Bb9H/CynCKwQK0w==,type:str]
```

Which...I think it would be better to encrypt the key as a single blob rather than encrypting each individual field, but it'll do the job, and I assume that the current production keys are encrypted similarly.

@jrgm @jbuck what do you think, is this a step in the right direction?

### CODE REVIEW

File: packages/fxa-auth-server/fxa-oauth-server/scripts/sops-key-config.js [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+const fs = require('fs');
+const assert = require('assert');
+const config = require('../lib/config');
+const yaml = require('js-yaml');
+const { execFileSync } = require('child_process');
+
+const SOPS_ITEM_KEY = 'openid.key';
+const SOPS_ITEM_NEWKEY = 'openid.newKey';
+const SOPS_ITEM_OLDKEY = 'openid.oldKey';
```
Comment (rfk) [inline packages/fxa-auth-server/fxa-oauth-server/scripts/sops-key-config.js]: I don't know what the actual tree paths of these config items are in our actual config file. @jbuck?

Comment (jbuck) [reply packages/fxa-auth-server/fxa-oauth-server/scripts/sops-key-config.js]: we're currently using `fxa_oauth::oauth_openid_key` for `SOPS_ITEM_KEY` - I guess we could use `fxa_oauth::oauth_openid_new_key` and `fxa_oauth::oauth_openid_old_key` for the equivalents?

**Comment by jbuck** [comment]
https://github.com/mozilla/fxa/pull/1668/files#diff-0adeac4fdea125ff2c6b3f1d1dc608c3R426 might need to be changed to the CWD of the process - resolving based on the `__dirname` was very confusing to me while I was working with it locally

**Comment by jbuck** [comment]
Okay, here's a list of steps to add a new key to fxa oauth stage after I patched the script to use my desired keys (`fxa_oauth::oauth_openid_key`, `fxa_oauth::oauth_openid_new_key`, and `fxa_oauth::oauth_openid_old_key`):

# One-time repo preparation
1. `git clone https://github.com/mozilla/fxa.git`
2. `cd fxa/packages/fxa-auth-server`
3. `npm install`
4. `cd fxa-oauth-server`
5. `echo '{}' | tee config/sopsKey.json config/sopsNewKey.json config/sopsOldKey.json`

# Generate new private key
1. `export FXA_OPENID_KEYFILE=../config/sopsKey.json`
2. `export FXA_OPENID_NEWKEYFILE=../config/sopsNewKey.json`
3. `export FXA_OPENID_OLDKEYFILE=../config/sopsOldKey.json`
4. `node scripts/sops-key-config.js extract ~/Documents/Github/sops/fxa.oauth.stage.yaml`
5. `node scripts/prepare-new-signing-key.js`
6. `node scripts/sops-key-config.js insert ~/Documents/Github/sops/fxa.oauth.stage.yaml`

@rfk Not sure if this would be considered a bug or not, but look how  `fxa_oauth::oauth_openid_new_key.fxa-createdAt` is serialized: `1.5644232e+09` . I'm not sure if this is a bug with the script, or with sops

**Comment by rfk** [comment]
> we're currently using fxa_oauth::oauth_openid_key for SOPS_ITEM_KEY - I guess we could
> use fxa_oauth::oauth_openid_new_key and fxa_oauth::oauth_openid_old_key for the equivalents?

Thanks, updated.

> Not sure if this would be considered a bug or not, but look how
> fxa_oauth::oauth_openid_new_key.fxa-createdAt is serialized: 1.5644232e+09

Huh. That's weird but also totally valid in both JSON and YAML, so I don't think it will cause any problems in practice.

> might need to be changed to the CWD of the process
> - resolving based on the __dirname was very confusing to me while I was working with it locally

Aye. What I think this is *intended* to do is to default to putting the keys into the `./config` directory, but it currently achieves it in a bit of a weird way; it sets the keyFile path to "../config/key.json" and then resolves it relative to the script.

@jbuck would it be useful to have the script calculate default values for these paths if they're not found in the loaded config?

> echo '{}' | tee config/sopsKey.json config/sopsNewKey.json config/sopsOldKey.json

I'm surprised that this was necessary; was it giving you an error if these files didn't exist?

**Comment by rfk** [comment]
I'm going to go ahead and merge this based on @jbuck's r+, we can do some followups to improve ease-of-use after trying it out in the wild.

**Pull request merged**